Nullable.md

```diff
@@ -0,0 +1,442 @@
+
+```
+function sum(xs::Vector{Union{Some{Int}, Void}})
+    s = 0
+    for x in xs
+        s += x
+    end
+    return x
+end
+```
+
+What type does this return when `xs = ?Int[]`? What type does it return when
+`xs = ?Int[Some(1)]`? What type does it return when `xs = ?Int[nothing]`?
+
+At a minimum, the use of `Some{T}` would require careful thought about
+type conversion and type promotion rules since we would need to define methods
+for combinations like `+(x::Int, y::Some{T})`. Almost surely, we would need to
+allow promoting `T` to `Some{T}` to make common code patterns work without
+much verbosity.
+
+### `Union{T, Null{T}}` Pros & Cons
+
+This final approach allows us to distinguish `Null{Bool}` from `Null{Int}` in
+cases in which that distinction might be useful. R makes such a distinction,
+although it seems to be done only because it is falls out naturally from an
+implementation that uses distinct sentinel values for each nullable type in
+the language. Otherwise, this approach is similar to the `Union{T, Void}` case
+described above, except for one crucial problem that would seem to suggest this
+approach offers few benefits over our current implementation of `Nullable{T}`.
+
+That problem is essentially the problem we already have with `Nullable{T}`:
+writing functions that return concrete types as output generally requires
+information about the return types of any functions used inside of a function
+because one needs to know the return type associated with the non-null case
+even when writing code to handle the null case. Needing to know about the code
+path not taken will be referred to the "counterfactual return type problem" in
+what follows.
+
+To see how this issue arises in practice, consider the following possible
+implementation of `map(f, x::Nullable{T})` given our current implementation of
+`Nullable{T}`:
+
+```
+function map(f, x::Nullable{T})
+    if !isnull(x)
+        return Nullable(f(get(x)))
+    else
+        return Nullable{Core.Inference.return_type(f, (T, ))}()
+    end
+end
+```
+
+For example, ensuring that `map(sqrt, Nullable{Int})` always returns
+`Nullable{Float64}` requires that the null path in the code be aware of the
+return type of `sqrt(x::Int)`. This information is not generally available in
+Julia.
+
+In static languages with arrow types, the counterfactual return type problem is
+not a problem at all because the return type of a function is guaranteed to be
+well-defined by the semantics of the language and guaranteed to be knowable at
+compile-time, so one can refer to properties of `f` without having to call into
+the implementation details of the language. But Julia does not provide such
+guarantees, so type-inference must be invoked to determine information about
+the return type of the counterfactual branch. An alternative approach is to
+use `promote_type` to explicitly register return types, but this is clearly
```

Comment:
`promote_op`? Anyway, I don't think it's worth mentioning as it's clearly not a solution.